fix(core): attempt to recover from user errors during creation

crisbeto opened the pull request:
If there's an error during the first creation pass of a `TView`, the data structure may be corrupted which will cause framework assertion failures downstream which can mask the user's error. These changes add a new flag to the `TView` that indicates whether the first creation pass was successful, and if it wasn't try re-create the `TView`.

Fixes #31221.

crisbeto commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -373,8 +372,13 @@ export function renderView<T>(tView: TView, lView: LView, context: T): void {
+    // Note that while we set this flag above, we need to do it here as well, because
+    // execution might not reach it if the template function throws an error.
+    if (tView.firstCreatePass) {
+      tView.firstCreatePass = false;
```
Note for reviewer: I was going back-and-forth between doing this and wrapping the template call above with a try/finally. I ended up doing this, because even if we do handle the error above, the query and `renderChildComponents` calls will probably be thrown as well, further masking the user's error.

crisbeto commented on integration/_payload-limits.json:
```diff
@@ -49,7 +49,7 @@
       "uncompressed": {
         "runtime-es2015": 2289,
-        "main-es2015": 226144,
+        "main-es2015": 226674,
```
I doubt this size increase is from my changes, master is most likely on the edge.

IgorMinar replied:
yeah, it's unrelated and already being updated in another PR. I can lgtm for this change, but you'll have to wait for AndrewK to review the rest.

IgorMinar reviewed (approved):
lgtm - but only for the unrelated size change

AndrewKushnir commented on packages/core/test/acceptance/view_insertion_spec.ts:
```diff
@@ -661,4 +661,44 @@ describe('view insertion', () => {
+
+  it('should not mask user errors during creation with framework errors', () => {
+    // Note that this use case is somewhat contrived, but it's inteded to
+    // mimic the behavior when using the router. See #31221 for more context.
```
Thanks for adding a comment @crisbeto! The logic in this component is a bit tricky (as you pointed out in the comment), i.e. we have the `<error-wrapper>` in a template and then we also create the same component via `ViewContainerRef.createComponent`, which is also wrapped in try-catch. I think it'd be great to expand the comment a bit to describe why we create the same component and why try-catch block is needed. Also it might make sense to add a reference to a function in Router that performs similar actions.

Otherwise, the changes look good to me 👍

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -344,7 +344,6 @@ export function renderView<T>(tView: TView, lView: LView, context: T): void {
     if (templateFn !== null) {
       executeTemplate(tView, lView, templateFn, RenderFlags.Create, context);
     }
-
```
nit: I think having an extra line here is consistent with other blocks below (that are also separated by empty lines), so I'd propose to revert this change.

AndrewKushnir reviewed (approved):
Thanks for fixing this problem and creating a test case for it (which was tricky)! 👍

The changes LGTM (with a couple nits), but I think it'd be great if @pkozlowski-opensource can have a look at this change as well, so I'm adding him as a reviewer to this PR.

Thank you.

crisbeto commented:
I've addressed the feedback @AndrewKushnir. Also sorry for the larger diff, `clang-format` ended up reformatting the files I changed.

crisbeto committed:
fix(core): attempt to recover from user errors during creation

If there's an error during the first creation pass of a `TView`, the data structure may be corrupted which will cause framework assertion failures downstream which can mask the user's error. These changes add a new flag to the `TView` that indicates whether the first creation pass was successful, and if it wasn't we try re-create the `TView`.

Fixes #31221.

crisbeto commented:
This is ready for review again.

crisbeto commented on packages/core/src/render3/interfaces/view.ts:
```diff
@@ -684,6 +684,12 @@ export interface TView {
+   * Indicates that there was an error before we managed to complete the first create pass of the
+   * view. This means that the view is likely corrupted and we should try to recover it.
+   */
+  incompleteFirstPass: boolean;
```
I was thinking of potentially making this a counter so that we can stop trying after a certain number of attempts. I decided to keep it simple for now, but I'm open to changing it.

pkozlowski-opensource replied:
Yeh, I think that keeping it simple for now is the way to go 👍

pkozlowski-opensource commented on packages/core/test/acceptance/view_insertion_spec.ts:
```diff
@@ -398,11 +398,11 @@ describe('view insertion', () => {
         fixture.detectChanges();
 
-        expect(fixture.debugElement.nativeElement.textContent).toBe('start|test|end');
+        expect(fixture.nativeElement.textContent).toBe('start|test|end');
```
👍

pkozlowski-opensource commented on packages/core/test/acceptance/view_insertion_spec.ts:
```diff
@@ -398,11 +398,11 @@ describe('view insertion', () => {
         fixture.componentInstance.insertTpl = true;
         fixture.detectChanges();
-        expect(fixture.debugElement.nativeElement.textContent).toBe('start|testtest|end');
+        expect(fixture.nativeElement.textContent).toBe('start|testtest|end');
```
👍

pkozlowski-opensource reviewed (approved):
LGTM.

AndrewKushnir commented:
[Presubmit](http://test/OCL:308835838:BASE:308835885:1588090658900:f874594d)

AndrewKushnir commented:
@crisbeto could you please rebase this PR? There is no conflict, but the merge script failed since this PR was rebased/created prior to the most recent update of the base SHA in the merge script. Feel free to add "merge" label back once it's done.

FYI, g3 presubmit is successful.

crisbeto commented:
Rebased.

AndrewKushnir commented:
@crisbeto thanks for rebasing! It looks like payload size limit checks failed on CI, could you please have a look when you get a chance? Thank you.

crisbeto committed:
fixup! fix(core): attempt to recover from user errors during creation

crisbeto commented:
Fixed the payload size failures.

AndrewKushnir commented:
Thanks @crisbeto.

I've re-requested review from @IgorMinar to review/approve payload size increase.

kara commented on goldens/size-tracking/integration-payloads.json:
```diff
@@ -30,7 +30,7 @@
       "uncompressed": {
         "runtime-es2015": 1485,
-        "main-es2015": 137320,
+        "main-es2015": 137897,
```
It looks like this is mostly accumulation from other PRs? The change itself doesn't seem to merit this jump in size.

kara reviewed (approved):
LGTM for size changes (look like accumulation)

IgorMinar reviewed (approved):
Lgtm! Thanks!
